### PR TITLE
#142 elastic implement complex fields

### PR DESCRIPTION
**Pull request by alfonso-noriega**

### CODE REVIEW

File: api/src/main/java/com/rbmhtechnology/vind/model/ComplexFieldDescriptorBuilder.java [hunk 1]
```diff
@@ -97,12 +97,15 @@
...
-                } else {
-                    return null;
+                if (Collection.class.isAssignableFrom(c.getClass())) {
+                    final Iterator iterator = ((Collection) c).iterator();
+                    if (iterator.hasNext()) {
+                        return storeFunction.apply((T) iterator.next());
+                    } else {
+                        return null;
+                    }
                 }
```
Comment (tkurz) [inline api/src/main/java/com/rbmhtechnology/vind/model/ComplexFieldDescriptorBuilder.java:107]: could be a else block

File: api/src/main/java/com/rbmhtechnology/vind/model/ComplexFieldDescriptorBuilder.java [hunk 2]
```diff
@@ -430,12 +433,15 @@
...
-                    return null;
+                if(Collection.class.isAssignableFrom(c.getClass())) {
+                    final Iterator iterator = ((Collection)c).iterator();
+                    if (iterator.hasNext()) {
+                        return storeFunction.apply((T)iterator.next());
+                    } else {
+                        return null;
+                    }
                 }
+                return storeFunction.apply((T)c);
```
Comment (tkurz) [inline api/src/main/java/com/rbmhtechnology/vind/model/ComplexFieldDescriptorBuilder.java:444]: else block

File: backend/elasticsearch/src/main/java/com/rbmhtechnology/vind/elasticsearch/backend/util/DocumentUtil.java [hunk 3]
```diff
@@ -63,6 +60,35 @@
...
+    private static void addFieldToDoc(Document doc, Map<String, Object> docMap, FieldDescriptor<?> descriptor) {
+        if (ComplexFieldDescriptor.class.isAssignableFrom(descriptor.getClass())) {
+            addFieldToDoc(doc, docMap, (ComplexFieldDescriptor) descriptor);
+        } else {
+            doc.getFieldContexts(descriptor)
+                    .forEach(context ->
+                            Optional.ofNullable(FieldUtil.getFieldName(descriptor, context))
+                                    .ifPresent(fieldName ->
+                                            Optional.ofNullable(doc.getContextualizedValue(descriptor, context))
+                                                    .ifPresent(value -> docMap.put(fieldName.replaceAll("\\.\\w+" , ""), toElasticType(value)))
```
Comment (tkurz) [inline backend/elasticsearch/src/main/java/com/rbmhtechnology/vind/elasticsearch/backend/util/DocumentUtil.java:72]: regex the way to go?

File: backend/elasticsearch/src/main/java/com/rbmhtechnology/vind/elasticsearch/backend/util/DocumentUtil.java [hunk 4]
```diff
@@ -92,6 +118,103 @@ private static Object toElasticType(Object value) {
         return value;
     }
 
+    private static Object toElasticType(Object value, ComplexFieldDescriptor descriptor, FieldUtil.Fieldname.UseCase useCase) {
```
Comment (tkurz) [inline backend/elasticsearch/src/main/java/com/rbmhtechnology/vind/elasticsearch/backend/util/DocumentUtil.java:121]: can you add some comments what you are doing here?

Comment (tkurz) [inline backend/elasticsearch/src/main/java/com/rbmhtechnology/vind/elasticsearch/backend/util/DocumentUtil.java:121]: you are working a lot with null values. Wouldn't it be nicer with optionals?

File: backend/elasticsearch/src/main/java/com/rbmhtechnology/vind/elasticsearch/backend/util/ResultUtils.java [hunk 5]
```diff
@@ -373,18 +373,22 @@ public static FacetResults buildFacetResults(Aggregations aggregations, Document
     }
 
     private static Pair<FieldDescriptor<?> ,TermFacetResult<?>> getTermFacetResults(Aggregation aggregation, Facet.TermFacet termFacet, DocumentFactory factory) {
+        return getTermFacetResults(aggregation, termFacet, factory, FieldUtil.Fieldname.UseCase.Facet);
+    }
+
+    private static Pair<FieldDescriptor<?> ,TermFacetResult<?>> getTermFacetResults(Aggregation aggregation, Facet.TermFacet termFacet, DocumentFactory factory, FieldUtil.Fieldname.UseCase useCase) {
```
Comment (tkurz) [inline backend/elasticsearch/src/main/java/com/rbmhtechnology/vind/elasticsearch/backend/util/ResultUtils.java:379]: describe complex methods with short comments what happens

File: backend/elasticsearch/src/main/java/com/rbmhtechnology/vind/elasticsearch/backend/util/DocumentUtil.java [hunk 6]
```diff
@@ -118,10 +241,12 @@ public static Document buildVindDoc( Map<String, Object> docMap , DocumentFactor
         docMap.keySet().stream()
                 .filter(name -> ! Arrays.asList(FieldUtil.ID, FieldUtil.TYPE, FieldUtil.SCORE, FieldUtil.DISTANCE)
                         .contains(name))
+                .filter(name ->  name.startsWith("complex_") && name.contains("_stored_") || name.startsWith("dynamic_"))
                 .forEach(name -> {
                     final Object o = docMap.get(name);
                     final String contextPrefix = searchContext != null ? searchContext + "_" : "";
-                    final Matcher internalPrefixMatcher = Pattern.compile(FieldUtil.INTERNAL_FIELD_PREFIX).matcher(name);
+                    final String pattern = "(" + FieldUtil.INTERNAL_FIELD_PREFIX + ")|(" + FieldUtil.INTERNAL_COMPLEX_FIELD_PREFIX + ")";
+                    final Matcher internalPrefixMatcher = Pattern.compile(pattern).matcher(name);
```
Comment (ja-fra) [inline backend/elasticsearch/src/main/java/com/rbmhtechnology/vind/elasticsearch/backend/util/DocumentUtil.java:249]: `Pattern.quote(...)`

File: backend/elasticsearch/src/main/java/com/rbmhtechnology/vind/elasticsearch/backend/util/DocumentUtil.java [hunk 7]
```diff
@@ -118,10 +241,12 @@ public static Document buildVindDoc( Map<String, Object> docMap , DocumentFactor
         docMap.keySet().stream()
                 .filter(name -> ! Arrays.asList(FieldUtil.ID, FieldUtil.TYPE, FieldUtil.SCORE, FieldUtil.DISTANCE)
                         .contains(name))
+                .filter(name ->  name.startsWith("complex_") && name.contains("_stored_") || name.startsWith("dynamic_"))
```
Comment (ja-fra) [inline backend/elasticsearch/src/main/java/com/rbmhtechnology/vind/elasticsearch/backend/util/DocumentUtil.java:244]: Use existing constants (`FieldUtil`)!

File: backend/elasticsearch/src/main/java/com/rbmhtechnology/vind/elasticsearch/backend/util/FieldUtil.java [hunk 8]
```diff
@@ -82,26 +98,108 @@ public static String getFieldName(FieldDescriptor<?> descriptor, String context)
...
+                return fieldName + type.getName() + contextPrefix + descriptor.getName();
+            }
+        }
     }
 
     public static String getSourceFieldName(String elasticFieldName, String context) {
         final String contextPrefix = context != null ? context + "_" : "";
-        final Matcher internalPrefixMatcher = Pattern.compile(FieldUtil.INTERNAL_FIELD_PREFIX).matcher(elasticFieldName);
+        final String pattern = "(" + FieldUtil.INTERNAL_FIELD_PREFIX + ")|(" + FieldUtil.INTERNAL_COMPLEX_FIELD_PREFIX + ")";
+        final Matcher internalPrefixMatcher = Pattern.compile(pattern).matcher(elasticFieldName);
```
Comment (ja-fra) [inline backend/elasticsearch/src/main/java/com/rbmhtechnology/vind/elasticsearch/backend/util/FieldUtil.java:197]: `Pattern.quote(...)`

**Pull request merged**